### ISSUE TITLE
`copyfile(3)` Replaces Data with a Broken HFS Compressed Resource Fork

### ISSUE THREAD

**Issue by hohle** [ttkb-oss/dedup#2]
Like #1 , `copyfile(3)` will produce unexpected results occasionally, but consistently. `dedup` uses `copyfile(3)` to transfer attributes, permissions, and ACLs from an original file to the clone that will replace it. The clone already has been created with `clonefile(2)` as a temporary file in the same directory as the target. After `copyfile` is called, the main data stream of the file (the one you expect to see during normal file operations) is truncated and the data appears to be compressed and moved into the resource fork. This appears to be standard transparent HFS compression, but however the resource fork and `com.apple.decmpfs` are corrupted in some way that the file is not recognized as a compressed file.

```
$> xattr -lx -p com.apple.decmpfs ~/Downloads/.~.file
com.apple.decmpfs:
00000000  66 70 6D 63 08 00 00 00 20 88 00 00 00 00 00 00  |fpmc.... .......|
00000010
```

The original file dos not use transparent compression and it is not expected that `copyfile(3)` will modify data at all.

There's nothing immediately obvious in [`copyfile.c`](https://opensource.apple.com/source/copyfile/copyfile-173.40.2/copyfile.c.auto.html) that would point to why this behavior exists. Some work probably needs to be done to trace these calls.

As a mitigation the replacement is not completed when this is detected. Following the call to `copyfile`, `replace_with_clone` will check the file size of the clone and fail if it is zero. Zero byte files are already ignored during candidate selection, so finding one at this point violates the invariant.

https://github.com/ttkb-oss/dedup/blob/93e1b962f094c33a987785ad14e60d14edb6f461/clone.c#L110C1-L118C6

**Comment by hohle** [comment]
The attached tarball contains a test driver and another tarball with test data.

After extracting the tarball, this reproduces the issue on my MacBook Pro M1 running Ventura 13.5.2 (22G91).

```bash
wget 'https://github.com/ttkb-oss/dedup/files/12709640/copyfile-bug.tar.gz' && \
    tar xvzf copyfile-bug.tar.gz && \
    mkdir -p tmp && \
    cc -o clonefile-bug clonefile-bug.c && \
    rm -rf tmp/clone-bug && \
    pushd tmp && \
    tar xvzf ../clone-bug.tar.gz && \
    popd && \
    ./clonefile-bug tmp/clone-bug/dog3_norm16.gim tmp/clone-bug/second/another.gim 
```

I'll test again after updating to 13.6.0 and try to find another system to test on prior to filing a bug with Apple.

[copyfile-bug.tar.gz](https://github.com/ttkb-oss/dedup/files/12709640/copyfile-bug.tar.gz)

**Comment by hohle** [comment]
Reproduced on an 2015 Intel MBA running 12.6.8 using the steps above.

**Comment by hohle** [comment]
Fixed in 8d8adf71f966ae8ab4aeccfbd6203d8caceecb79
